### PR TITLE
[Model] Support math-shepherd-mistral-7b-prm model

### PR DESCRIPTION
**Pull request by Went-Liang**
FILL IN THE PR DESCRIPTION HERE

Support [peiyi9979/math-shepherd-mistral-7b-prm](https://huggingface.co/peiyi9979/math-shepherd-mistral-7b-prm) as embedding model.

As mentioned by [9314](https://github.com/vllm-project/vllm/issues/9314), the Process-Supervised Reward Model, which provides reward scores for intermediate steps generated by LLMs, can offer more fine-grained optimization for Reinforcement Learning (RL). This will help the community reproduce the OpenAI O1 model. PR [9424](https://github.com/vllm-project/vllm/pull/9424) allows any model that adds a pooler method to be used as an embedding model. 

Therefore, this PR adds a pooler method to `LlamaForCausalLM`, introduces a `pooling-type` named "STEP" and adds a `PoolerConfig` class to facilitate users to configure the pooler method. In STEP mode, users can use the `peiyi9979/math-shepherd-mistral-7b-prm` model by setting the `pooling-step-tag-id` and `pooling-returned-token-ids` variables. `pooling-returned-token-ids` represents a list of indices for the vocabulary dimensions to be extracted, such as the token IDs of `good_token` and `bad_token` in the [math-shepherd-mistral-7b-prm model](https://huggingface.co/peiyi9979/math-shepherd-mistral-7b-prm). When `pooling-step-tag-id` is not None, it indicates that the score corresponding to the `pooling-step-tag-id` in the generated sentence should be returned. Otherwise, it returns the scores for all tokens.

The model can be served with:
```python
python -m vllm.entrypoints.openai.api_server \
	--model peiyi9979/math-shepherd-mistral-7b-prm \
	--trust-remote-code \
	--served-model-name math-shepherd-mistral-7b-prm \
	--port 8080 \
	--tensor-parallel-size 1 \
	--enforce-eager \
        --task embedding \
	--pooling-type STEP \
	--pooling-step-tag-id 12902 \
	--pooling-returned-token-ids 648 387
```

And a test correspond to the example in the huggingface model page is:
```python
import torch
from openai import OpenAI

question = """Janet\u2019s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?"""
output1 = """Step 1: Janet's ducks lay 16 eggs per day. ки\nStep 2: She eats three for breakfast every morning, so she has 16 - 3 = 13 eggs left. ки\nStep 3: She bakes muffins for her friends every day with four eggs, so she has 13 - 4 = 9 eggs left. ки\nStep 4: She sells the remainder at the farmers' market daily for $2 per fresh duck egg, so she makes 9 * $2 = $18 every day at the farmers' market. The answer is: 18 ки""" # 18 is right
output2 = """Step 1: Janet's ducks lay 16 eggs per day. ки\nStep 2: She eats three for breakfast every morning, so she has 16 - 3 = 13 eggs left. ки\nStep 3: She bakes muffins for her friends every day with four eggs, so she has 13 - 4 = 9 eggs left. ки\nStep 4: She sells the remainder at the farmers' market daily for $2 per fresh duck egg, so she makes 9 * $2 = $17 every day at the farmers' market. The answer is: 17 ки""" # 17 is wrong
full_prompt = [f"{question} {output}" for output in [output1, output2]]


# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8080/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

models = client.models.list()
model = models.data[0].id

responses = client.embeddings.create(
    input=full_prompt,
    model=model,
)

for data in responses.data:
    print(torch.tensor(data.embedding).view(-1, 2)[:, 0])
# tensor([0.9956, 0.9956, 0.9985, 0.9956])
# tensor([0.9956, 0.9956, 0.9985, 0.0240])
```

Of course, you can also use it directly like this:
```python
from vllm import LLM

llm = LLM(
    model='peiyi9979/math-shepherd-mistral-7b-prm', 
    tensor_parallel_size=1, 
    task="embedding",
    pooling_type="STEP",
    pooling_step_tag_id=12902,
    pooling_returned_token_ids=[648, 387],
)

question = """Janet\u2019s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?"""
output1 = """Step 1: Janet's ducks lay 16 eggs per day. ки\nStep 2: She eats three for breakfast every morning, so she has 16 - 3 = 13 eggs left. ки\nStep 3: She bakes muffins for her friends every day with four eggs, so she has 13 - 4 = 9 eggs left. ки\nStep 4: She sells the remainder at the farmers' market daily for $2 per fresh duck egg, so she makes 9 * $2 = $18 every day at the farmers' market. The answer is: 18 ки""" # 18 is right
output2 = """Step 1: Janet's ducks lay 16 eggs per day. ки\nStep 2: She eats three for breakfast every morning, so she has 16 - 3 = 13 eggs left. ки\nStep 3: She bakes muffins for her friends every day with four eggs, so she has 13 - 4 = 9 eggs left. ки\nStep 4: She sells the remainder at the farmers' market daily for $2 per fresh duck egg, so she makes 9 * $2 = $17 every day at the farmers' market. The answer is: 17 ки""" # 17 is wrong

good_token = '+'
bad_token = '-'
step_tag = 'ки'

tokenizer = llm.get_tokenizer()
[good_token_id, bad_token_id, step_tag_id] = tokenizer.encode(f"{good_token} {bad_token} {step_tag}")[1:] # [648, 387, 12902]


full_prompt = [f"{question} {output}" for output in [output1, output2]]
outputs = llm.encode(full_prompt)
```

Thank you for your time on reviewing this PR :)

### CODE REVIEW

File: vllm/model_executor/layers/pooler.py [hunk 1]
```diff
@@ -33,6 +35,9 @@ def __init__(self, pooling_type: PoolingType, normalize: bool):
 
         self.pooling_type = pooling_type
         self.normalize = normalize
+        returned_token_ids = os.environ.get('RETURNED_TOKEN_IDS', '648,387')
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:38]: Setting the environment variable may work as a quick hack, but it's not really maintainable. I suggest adding a new config `PoolerConfig` to pass this information via `EngineArgs` to the model (similar to how we have `MultiModalConfig`), which can be used to initialize the `Pooler` instance.

Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:38]: Also see #9506

Comment (Went-Liang) [reply vllm/model_executor/layers/pooler.py]: Thank you very much for your suggestions. I have made the changes as you recommended.

File: vllm/config.py [hunk 2]
```diff
@@ -242,6 +269,19 @@ def _init_multimodal_config(
 
         return None
 
+    def _init_pooler_config(
+            self, pooling_type, pooling_norm, pooling_softmax,
+            pooling_step_tag_id,
+            pooling_returned_token_ids) -> Optional["PoolerConfig"]:
```
Comment (DarkLight1337) [inline vllm/config.py:275]: Can you add type annotations to this method?

File: vllm/engine/arg_utils.py [hunk 3]
```diff
@@ -850,6 +857,42 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
...
+        parser.add_argument(
+            '--pooling-returned-token-ids',
+            nargs='+',
+            type=int,
+            default=None,
+            help="pooling-returned-token-ids represents a list of "
+            "indices for the vocabulary dimensions to be extracted, "
+            "such as the token IDs of good_token and bad_token in "
+            "the math-shepherd-mistral-7b-prm model.")
+
```
Comment (DarkLight1337) [inline vllm/engine/arg_utils.py:895]: The default should be based on the model. Can we make `None` the default for these arguments?

File: vllm/config.py (hunk 2 above)
Comment (Went-Liang) [reply vllm/config.py]: I have made the changes as you recommended.

File: vllm/engine/arg_utils.py (hunk 3 above)
Comment (Went-Liang) [reply vllm/engine/arg_utils.py]: I have made the changes as you recommended.

Comment (DarkLight1337) [reply vllm/engine/arg_utils.py]: The boolean arguments should also be None by default (to distinguish them from False for models that enable them by default)

Comment (Went-Liang) [reply vllm/engine/arg_utils.py]: Please check it.

Comment (DarkLight1337) [reply vllm/engine/arg_utils.py]: Please update the type annotation accordingly.

Comment (DarkLight1337) [reply vllm/engine/arg_utils.py]: Also, you have to update each embedding model to use the PoolerConfig.

Comment (Went-Liang) [reply vllm/engine/arg_utils.py]: I have modified it according to your suggestion, please check it

File: vllm/model_executor/models/bert.py [hunk 4]
```diff
@@ -387,10 +387,18 @@ def __init__(
...
         self.model = BertModel(config, cache_config, quant_config)
-        self._pooler = Pooler(pooling_type=PoolingType.CLS, normalize=True)
+        self._pooler = Pooler(
+            pooling_type=PoolingType[pooler_config.pooling_type]
+            if pooler_config.pooling_type is not None else PoolingType.CLS,
+            normalize=pooler_config.pooling_norm or True,
+            softmax=pooler_config.pooling_softmax or False,
+            step_tag_id=pooler_config.pooling_step_tag_id,
+            returned_token_ids=pooler_config.pooling_returned_token_ids,
+        )
```
Comment (DarkLight1337) [inline vllm/model_executor/models/bert.py:401]: Can we add a factory method to `Pooler` so automatically merge the config with model specific defaults?

Comment (DarkLight1337) [reply vllm/model_executor/models/bert.py:401]: e.g. we should be able to write

```
self._pooler = Pooler.from_config_with_defaults(
    pooler_config,
    default_pooling_type=PoolingType.CLS,
    default_normalize=True,
)
```

Comment (Went-Liang) [reply vllm/model_executor/models/bert.py]: Please check it

File: vllm/model_executor/layers/pooler.py [hunk 5]
```diff
@@ -28,15 +31,41 @@ class Pooler(nn.Module):
...
+        pooling_type: Optional[PoolingType] = None,
+        normalize: Optional[bool] = None,
+        softmax: Optional[bool] = None,
+        step_tag_id: Optional[int] = None,
+        returned_token_ids: Optional[List[int]] = None,
+    ) -> "Pooler":
+        return cls(
+            pooling_type=PoolingType[pooler_config.pooling_type]
+            if pooler_config.pooling_type is not None else pooling_type,
+            normalize=pooler_config.pooling_norm or normalize,
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:63]: We should check `None` explicitly. Otherwise, the model default will override an explicitly-set `False` value.

File: vllm/model_executor/layers/pooler.py [hunk 6]
```diff
@@ -28,15 +31,41 @@ class Pooler(nn.Module):
...
+            softmax=pooler_config.pooling_softmax or softmax,
+            step_tag_id=pooler_config.pooling_step_tag_id or step_tag_id,
+            returned_token_ids=pooler_config.pooling_returned_token_ids
+            or returned_token_ids,
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:67]: We should check `None` explicitly. Otherwise, the model default will override an explicitly-set `False` value.

File: vllm/model_executor/layers/pooler.py [hunk 7]
```diff
@@ -28,15 +31,41 @@ class Pooler(nn.Module):
...
-                 pooling_type: PoolingType,
-                 normalize: bool,
-                 softmax: bool = False):
+    def __init__(
+        self,
+        pooling_type: Optional[PoolingType] = None,
+        normalize: Optional[bool] = None,
+        softmax: Optional[bool] = None,
+        step_tag_id: Optional[int] = None,
+        returned_token_ids: Optional[List[int]] = None,
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:40]: Meanwhile here the original arguments (pooling type, normalize, softmax) should not be `None` as they have already been overridden.

File: vllm/model_executor/layers/pooler.py [hunk 8]
```diff
@@ -28,15 +31,41 @@ class Pooler(nn.Module):
...
 
-    def __init__(self,
-                 pooling_type: PoolingType,
-                 normalize: bool,
-                 softmax: bool = False):
+    def __init__(
+        self,
+        pooling_type: Optional[PoolingType] = None,
+        normalize: Optional[bool] = None,
+        softmax: Optional[bool] = None,
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:38]:
```suggestion
        pooling_type: PoolingType,
        normalize: bool,
        softmax: bool,
```

The original arguments (pooling type, normalize, softmax) should not be `None` as they have already been overridden.

File: vllm/model_executor/layers/pooler.py [hunk 9]
```diff
@@ -28,15 +31,41 @@ class Pooler(nn.Module):
...
+        self.step_tag_id = step_tag_id
+        self.returned_token_ids = returned_token_ids
+
+    @classmethod
+    def from_config_with_defaults(
+        cls,
+        pooler_config: PoolerConfig,
+        pooling_type: Optional[PoolingType] = None,
+        normalize: Optional[bool] = None,
+        softmax: Optional[bool] = None,
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:56]:
```suggestion
        pooling_type: PoolingType,
        normalize: bool,
        softmax: bool,
```

Same as constructor

Comment (Went-Liang) [reply vllm/model_executor/layers/pooler.py]: Please check it

**Review by DarkLight1337 — APPROVED** [review]
Looks good now, thanks for your effort and patience!

**Comment by Went-Liang** [comment]
> Now you just have to get the tests to pass.

Excuse me, the test produced the following error (as shown in the image). This doesn't seem to be caused by my code changes. Could you please advise on how to handle this?

image

**Comment by DarkLight1337** [comment]
I have retried the failing test, see if it passes this time.

**Comment by DarkLight1337** [comment]
Looks like this issue comes from main branch, I have asked those with permissions to force-merge this.
